test-data/unit/check-selftype.test

```diff
@@ -468,16 +468,225 @@ class B(A[Q]):
+reveal_type(ci.from_item)  # N: Revealed type is 'def (item: Tuple[builtins.int])'
+reveal_type(cs.from_item)  # N: Revealed type is 'def (item: builtins.str)'
+
+[case testSelfTypeRestrictedMethodOverloadFallback]
+from typing import TypeVar, Generic, overload, Callable
+
+T = TypeVar('T')
+class C(Generic[T]):
+    @overload
+    def from_item(self: C[str]) -> str: ...
+    @overload
+    def from_item(self, converter: Callable[[T], str]) -> str: ...
+    def from_item(self, converter):
+        ...
+
+ci: C[int]
+cs: C[str]
+reveal_type(cs.from_item())  # N: Revealed type is 'builtins.str'
+ci.from_item()  # E: Too few arguments for "from_item" of "C"
+
+def conv(x: int) -> str: ...
+def bad(x: str) -> str: ...
+reveal_type(ci.from_item(conv))  # N: Revealed type is 'builtins.str'
+ci.from_item(bad)  # E: Argument 1 to "from_item" of "C" has incompatible type "Callable[[str], str]"; expected "Callable[[int], str]"
+
+[case testSelfTypeRestrictedMethodOverloadInit]
+from lib import P, C
+
+reveal_type(P)  # N: Revealed type is 'Overload(def [T] (use_str: Literal[True]) -> lib.P[builtins.str], def [T] (use_str: Literal[False]) -> lib.P[builtins.int])'
+reveal_type(P(use_str=True))  # N: Revealed type is 'lib.P[builtins.str]'
+reveal_type(P(use_str=False))  # N: Revealed type is 'lib.P[builtins.int]'
+
+reveal_type(C)  # N: Revealed type is 'Overload(def [T] (item: T`1, use_tuple: Literal[False]) -> lib.C[T`1], def [T] (item: T`1, use_tuple: Literal[True]) -> lib.C[builtins.tuple[T`1]])'
+reveal_type(C(0, use_tuple=False))  # N: Revealed type is 'lib.C[builtins.int*]'
+reveal_type(C(0, use_tuple=True))  # N: Revealed type is 'lib.C[builtins.tuple[builtins.int*]]'
+
+[file lib.pyi]
+from typing import TypeVar, Generic, overload, Tuple
+from typing_extensions import Literal
+
+T = TypeVar('T')
+class P(Generic[T]):
+    @overload
+    def __init__(self: P[str], use_str: Literal[True]) -> None: ...
+    @overload
+    def __init__(self: P[int], use_str: Literal[False]) -> None: ...
+
+class C(Generic[T]):
+    @overload
+    def __init__(self: C[T], item: T, use_tuple: Literal[False]) -> None: ...
+    @overload
+    def __init__(self: C[Tuple[T, ...]], item: T, use_tuple: Literal[True]) -> None: ...
+
+[case testSelfTypeNarrowBinding]
+from typing import TypeVar, List, Generic
+
+T = TypeVar('T')
+S = TypeVar('S')
+
+class Base(Generic[T]):
+    def get_item(self: Base[List[S]]) -> S: ...
+
+class Sub(Base[List[int]]): ...
+class BadSub(Base[int]): ...
+
+reveal_type(Sub().get_item())  # N: Revealed type is 'builtins.int'
+BadSub().get_item()  # E: Invalid self argument "BadSub" to attribute function "get_item" with type "Callable[[Base[List[S]]], S]"
+[builtins fixtures/list.pyi]
+
+[case testMixinAllowedWithProtocol]
+from typing import TypeVar
+from typing_extensions import Protocol
+
+class Resource(Protocol):
+    def close(self) -> int: ...
+
+class AtomicClose:
+    def atomic_close(self: Resource) -> int:
+        return self.close()
```

Comment:
What happens if we add a second method to AtomicClose and try calling it inside here?

Do we end up treating the type of `self` as the intersection of AtomicClose and Resource and allow that to type-check? Or do we assume that `self` is just of type Resource and report an error?

Reply:
> Do we end up treating the type of `self` as the intersection of AtomicClose and Resource and allow that to type-check?

No, I mentioned in one of the issues that I don't really like the implicit intersections. I think if there is an explicit annotation, it should be just it. Later when we add intersections people will be able to put an explicit intersection there. Currently one can just add a minimal bunch of needed methods to the protocol.